torch/distributed/checkpoint/default_planner.py

```diff
@@ -169,12 +202,16 @@ def __init__(
         flatten_state_dict: bool = True,
         flatten_sharded_tensors: bool = True,
         allow_partial_load: bool = False,
+        extension_registry: Optional[ExtensionRegistry] = None,
```

Comment:
Just flagging one of the largest complaints from our users is that the planner/writer abstraction is too complicated and hard to understand what's going on. I wonder if the planner is the right place for this kind of transformation and whether we've explored doing this somewhere else?

Reply:
I think it's hard to understand because there aren't clear delineations of responsibility.  See the previous comment about serialization not being in a single place.  This is absolutely an area where DCP can improve.  I tried doing some cleanup but ran into some difficulties, so I shelved those cleanups in favor of solving the specific compression problem to start.

More generally, though, it seems valuable to reuse common code.  For both serialization and stream transforms, it seems preferable to provide a common implementation which all reader/writers can use.  Otherwise, improvements in these areas need to be made once per implementation, which is made worse by the possibility of local implementations which we don't control.  By using planner hooks which the reader/writer can call into, we give ourselves code sharing and flexibility.  That's why the code is structured the way it is.  We don't want to add compression separately to each reader/writer.

The pattern we should (I believe) encourage is for the reader/writer to specify the storage mechanism (filesystem, fsspec, S3, manifold, etc), and for the implementation to call into the planner for common logic (sharding, serialization, compression, batching, etc).  This makes adding new storage mechanisms easier, and means many improvements can be made centrally, but still gives flexibility to override the common patterns should a storage developer want to (I'm not sure why they would).

That said, you've correctly identified one of the big directional areas in this stack, and I'm happy to discuss it further if you think this is the wrong tradeoff.